testcontainers/testcontainers-go
# fix: enhance host configuration port binding

## What does this PR do?

This PR refactors the mergePortBinding function to accommodate exposed ports with their respective protocols.

## Why is it important?

It enables to configure port binding to an exposed port declared with protocol.

## Related issues

- Closes #2483

## How to test this PR

Create a container with following request should map exposed container redis port to port binding configured.

```go
genericContainerReq := testcontainers.GenericContainerRequest{
	ContainerRequest: req := testcontainers.ContainerRequest{
		Image:        "redis:6.2",
		Name:         "redis",
		ExposedPorts: []string{"6379/tcp"},
		WaitingFor:   wait.ForExposedPort(),
		HostConfigModifier: func(hc *container.HostConfig) {
			hc.PortBindings = map[nat.Port][]nat.PortBinding{
				"6379/tcp": {{HostIP: "0.0.0.0", HostPort: "6379"}},
			}
		},
	},
	Started:          true,
}

container, err := testcontainers.GenericContainer(context.Background(), genericContainerReq)
````

The container should be configured with port map 6379:6379

```sh
docker ps

CONTAINER ID   IMAGE                       COMMAND                  CREATED         STATUS         PORTS                     NAMES
ee95a18b3ccc   redis:6.2                   "docker-entrypoint.s…"   6 seconds ago   Up 6 seconds   0.0.0.0:6379->6379/tcp    redis
